src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp

```diff
@@ -0,0 +1,175 @@
+
+        const __m128 _x = _mm_load1_ps(px + i);
+        const __m128 _xx = _mm_mul_ps(_x, _x);
+
+        // tmp -= <v_j,f, v_j,f> * x * x
+        const int vBias = j * m * d + f * d;
+
+        // j-th feature's latent vector in the f-th field hidden space.
+        const float * vjf = pv + vBias;
+
+        for (int k = 0; k + 4 <= d; k += 4)
+        {
+            const __m128 _v = _mm_load_ps(vjf + k);
+            _tmp = _mm_sub_ps(_tmp, _mm_mul_ps(_mm_mul_ps(_v, _v), _xx));
+        }
+
+        for (int fprime = 0; fprime < m; fprime++)
+        {
+            const int vBias = j * m * d + fprime * d;
+            const int qBias = f * m * d + fprime * d;
+            const float * vjfprime = pv + vBias;
+            float * qffprime = pq + qBias;
+
+            // q_f,f' += v_j,f' * x
+            for (int k = 0; k + 4 <= d; k += 4)
+            {
+                const __m128 _v = _mm_load_ps(vjfprime + k);
+                __m128 _q = _mm_load_ps(qffprime + k);
+                _q = _mm_add_ps(_q, _mm_mul_ps(_v, _x));
+                _mm_store_ps(qffprime + k, _q);
+            }
+        }
+    }
+
+    for (int f = 0; f < m; f++)
+    {
+        // tmp += <q_f,f, q_f,f>
+        const float * qff = pq + f * m * d + f * d;
+        for (int k = 0; k + 4 <= d; k += 4)
+        {
+            __m128 _qff = _mm_load_ps(qff + k);
+
+            // Intra-field interactions. 
+            _tmp = _mm_add_ps(_tmp, _mm_mul_ps(_qff, _qff));
+        }
+
+        // y += <q_f,f', q_f',f>, f != f'
+        // Whis loop handles inter - field interactions because f != f'.
+        for (int fprime = f + 1; fprime < m; fprime++)
+        {
+            const float * qffprime = pq + f * m * d + fprime * d;
+            const float * qfprimef = pq + fprime * m * d + f * d;
+            for (int k = 0; k + 4 <= d; k += 4)
+            {
+                // inter-field interaction
+                __m128 _qffprime = _mm_load_ps(qffprime + k);
+                __m128 _qfprimef = _mm_load_ps(qfprimef + k);
+                _y = _mm_add_ps(_y, _mm_mul_ps(_qffprime, _qfprimef));
+            }
+        }
+    }
+
+    _y = _mm_add_ps(_y, _mm_mul_ps(_mm_set_ps1(0.5f), _tmp));
+    _tmp = _mm_add_ps(_y, _mm_movehl_ps(_y, _y));
+    _y = _mm_add_ps(_tmp, _mm_shuffle_ps(_tmp, _tmp, 1)); // The lowest slot is the response value.
+    _mm_store_ss(&latentResponse, _y);
+    *response = linearResponse + latentResponse;
+}
+
+// Calculate the stochastic gradient and update the model. 
+// The /*const*/ comment on the parameters of the function means that their values should not get altered by this function.
+EXPORT_API(void) CalculateGradientAndUpdateNative(float lambdaLinear, float lambdaLatent, float learningRate, int fieldCount, int latentDim, float weight, int count,
+    _In_ int* /*const*/ fieldIndices, _In_ int* /*const*/ featureIndices, _In_ float* /*const*/ featureValues, _In_ float* /*const*/ latentSum, float slope,
```

Comment:
Is it possible to actually enforce this constness instead of just via comment but either as  & reference or const?

Reply:
Thanks for suggesting.  Will take a look at the next iteration on this file, right after enabling the tests. 

---
In reply to: [197935966](https://github.com/dotnet/machinelearning/pull/383#discussion_r197935966) [](ancestors = 197935966)